### PR TITLE
First step of performance improvements

### PR DESCRIPTION
**Pull request by jkiesele**
A few lines to make each job faster by about a factor of 8.

### CODE REVIEW

File: python/fill_histos.py [hunk 1]
```diff
@@ -128,14 +129,13 @@ def fillhistos(year, region, dataset, systematic, number, cuts):
...
+            # the ntuple at this step!!!
+            scale = 1.
             if datasets[dataset]['MC']:
                 scale = 1000 * datasets[dataset]['XS'] * datasets[dataset][year]['KFactor'] * lumi[year]
-                histos[histname].Scale(scale)
-
-                print(f'selected {histos[histname].GetEntries()} events out of {datasetInfo["genEventCount"]} (genEventCount)')
-                print(f"scaled with {scale:.3g} = {1000*datasets[dataset]['XS']:.1f}(XS in fb) \
-    * {datasets[dataset][year]['KFactor']}(K-factor) * {lumi[year]}(lumi in 1/fb)")
+            scales[histname] = scale
```
Comment (marco-link) [inline python/fill_histos.py]: The scale only depends on the dataset and year (not on the histogram).
Can't we then just move the scaling to outside the histogram loop?

File: python/fill_histos.py [hunk 2]
```diff
@@ -128,14 +129,13 @@ def fillhistos(year, region, dataset, systematic, number, cuts):
...
+            # determine scales
+            # do not ask for the histogram entries or anything that requires processing
+            # the ntuple at this step!!!
+            scale = 1.
             if datasets[dataset]['MC']:
                 scale = 1000 * datasets[dataset]['XS'] * datasets[dataset][year]['KFactor'] * lumi[year]
-                histos[histname].Scale(scale)
-
-                print(f'selected {histos[histname].GetEntries()} events out of {datasetInfo["genEventCount"]} (genEventCount)')
-                print(f"scaled with {scale:.3g} = {1000*datasets[dataset]['XS']:.1f}(XS in fb) \
```
Comment (marco-link) [inline python/fill_histos.py:137]: I'd prefer to keep the scale printout for debugging.
Just move to the actual scaling part.
Is there a way to keep the event selection efficiency without affecting performance?

**Review by marco-link — CHANGES_REQUESTED** [review]

File: python/fill_histos.py (hunk 2 above)
Comment (jkiesele) [reply python/fill_histos.py:137]: - we can print the scale, sure.
- how can the selection efficiency be computed without running the event loop? That's the issue. We can print it somewhere below outside of the loop, e.g. where also the histogram scaling is applied

File: python/fill_histos.py (hunk 1 above)
Comment (jkiesele) [reply python/fill_histos.py]: that is precisely what I did. I just kept the structure parallel to the histogram dict.
Is the scale also constant if you consider different regions and systematics in parallel? 
(remember that there is the plan to restructure, which is why I attached it to a histogram to be safe here)

Comment (marco-link) [reply python/fill_histos.py]: Yes, the scale is constant considering different regions and systematics.
I think we want to restructure in a way that only one jobs runs over each individual file.
In that case the scale is the same for all histograms as it only depends on the dataset and year.

Comment (defranchis) [reply python/fill_histos.py]: Hi, just one small addition. The overall scale is the same for all regions, but not for all systematics. For example in the case of PDFs and ME variations we will have to adjust the overall scaling too

Comment (marco-link) [reply python/fill_histos.py]: @defranchis Don't we use weightsums for that or is there an additional factor?

Comment (defranchis) [reply python/fill_histos.py]: yes we use the sum of the weights, what you implemented in https://github.com/WbWbX/extraction/pull/5

Comment (marco-link) [reply python/fill_histos.py]: They are applied at a different place. This scale is solely to scale with the luminosity and cross section.

Comment (defranchis) [reply python/fill_histos.py]: ah OK, sorry for the confusion then

**Pull request merged**